algorithmic_efficiency/workloads/criteo1tb/criteo1tb_pytorch/workload.py

```diff
@@ -0,0 +1,177 @@
+      normalization = mask_batch.sum()
+    else:
+      weighted_losses = per_example_losses
+      normalization = label_batch.shape[0]
+    return torch.sum(weighted_losses, dim=-1) / normalization
+
+  def _eval_metric(self, logits: spec.Tensor,
+                   targets: spec.Tensor) -> Dict[str, int]:
+    loss = self.loss_fn(logits, targets).sum()
+    return {'loss': loss}
+
+  def init_model_fn(self, rng: spec.RandomState) -> spec.ModelInitState:
+    torch.random.manual_seed(rng[0])
+    model = DlrmSmall(
+        vocab_sizes=self.vocab_sizes,
+        total_vocab_sizes=sum(self.vocab_sizes),
+        num_dense_features=self.num_dense_features,
+        mlp_bottom_dims=(128, 128),
+        mlp_top_dims=(256, 128, 1),
+        embed_dim=64)
+
+    self._param_shapes = {
+        k: spec.ShapeTuple(v.shape) for k, v in model.named_parameters()
+    }
+    model.to(DEVICE)
+    if N_GPUS > 1:
+      if USE_PYTORCH_DDP:
+        model = DDP(model, device_ids=[RANK], output_device=RANK)
+      else:
+        model = torch.nn.DataParallel(model)
+    return model, None
+
+  def model_fn(
+      self,
+      params: spec.ParameterContainer,
+      augmented_and_preprocessed_input_batch: Dict[str, spec.Tensor],
+      model_state: spec.ModelAuxiliaryState,
+      mode: spec.ForwardPassMode,
+      rng: spec.RandomState,
+      update_batch_norm: bool) -> Tuple[spec.Tensor, spec.ModelAuxiliaryState]:
+    del model_state
+    del rng
+    del update_batch_norm
+
+    model = params
+    inputs = augmented_and_preprocessed_input_batch['inputs']
+
+    if mode == spec.ForwardPassMode.EVAL:
+      model.eval()
+
+    if mode == spec.ForwardPassMode.TRAIN:
+      model.train()
+
+    contexts = {
+        spec.ForwardPassMode.EVAL: torch.no_grad,
+        spec.ForwardPassMode.TRAIN: contextlib.nullcontext
+    }
+
+    with contexts[mode]():
+      logits_batch = model(inputs)
+
+    return logits_batch, None
+
+  def build_input_queue(self,
+                        data_rng: jax.random.PRNGKey,
+                        split: str,
+                        data_dir: str,
+                        global_batch_size: int,
+                        num_batches: Optional[int] = None,
+                        repeat_final_dataset: bool = False):
+    per_device_batch_size = int(global_batch_size / N_GPUS)
+
+    # The input pipeline has to be created in all processes, because
+    # self._tokenizer has to be available in every process.
+    np_iter = super().build_input_queue(data_rng,
+                                        split,
+                                        data_dir,
+                                        global_batch_size,
+                                        num_batches,
+                                        repeat_final_dataset)
+    while True:
+      # Only iterate over tf input pipeline in one Python process to
+      # avoid creating too many threads.
+      if RANK == 0:
```

Comment:
@runame It would be helpful if you can take a look at this part; I borrowed the code from WMT (with the modification that I think is needed). To give you more background, we have 'inputs', 'targets', and 'weights' in the batch and the dimension would be `batch_size, 39`. Thanks a lot!